File: mac/Visualizer/Visualizer/Renderer.swift
import Foundation
import Metal
import MetalKit

class Renderer: NSObject, MTKViewDelegate {

    let device: MTLDevice
    let commandQueue: MTLCommandQueue
    let pipelineState: MTLRenderPipelineState
    let vertexBuffer: MTLBuffer
    var vertices: [Vertex]

    init(view: MTKView) {
        self.device = view.device!
        guard let commandQueue = device.makeCommandQueue() else {
            fatalError("Could not make commandQueue")
        }
        self.commandQueue = commandQueue
        do {
            pipelineState = try Renderer.buildRenderPipelineWith(device: device, metalKitView: view)
        }
        catch {
            fatalError("\(error)")
        }

        vertices = [Vertex(color: [1, 0, 0, 1], pos: [-1, -1]),
                        Vertex(color: [0, 1, 0, 1], pos: [0, 1]),
                        Vertex(color: [0, 0, 1, 1], pos: [1, -1])]
        guard let buffer = device.makeBuffer(bytes: vertices, length: vertices.count * MemoryLayout<Vertex>.stride, options: []) else {
            fatalError("Could not make buffer")
        }
        vertexBuffer = buffer
    }

    class func buildRenderPipelineWith(device: MTLDevice, metalKitView: MTKView) throws -> MTLRenderPipelineState {
        let pipelineDescriptor = MTLRenderPipelineDescriptor()
        guard let library = device.makeDefaultLibrary() else {
            fatalError("Could not makeDefaultLibrary")
        }
        pipelineDescriptor.vertexFunction = library.makeFunction(name: "vertexShader")
        pipelineDescriptor.fragmentFunction = library.makeFunction(name: "fragmentShader")
        pipelineDescriptor.colorAttachments[0].pixelFormat = metalKitView.colorPixelFormat
        return try device.makeRenderPipelineState(descriptor: pipelineDescriptor)
    }

    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {

    }

    func draw(in view: MTKView) {

        for ii in 0..<vertices.count {
            var vertex = vertices[ii]
            print("hello \(vertex.pos)")
            var x = vertex.pos.x
            var y = vertex.pos.y
            if x <= -1 {
                x = -1
                if y < 1 {
                    y += 0.01
                }
                else {
                    x += 0.01
                }
            }
            else if y >= 1 {
                y = 1
                if x < 1 {
                    x += 0.01
                }
                else {
                    y -= 0.01
                }
            }
            else if x >= 1 {
                x = 1
                if y > -1 {
                    y -= 0.01
                }
                else {
                    x -= 0.01
                }
            }
            else if y <= -1 {
                y = -1
                if x > -1 {
                    x -= 0.01
                }
                else {
                    y += 0.01
                }
            }

            vertex.pos = [x, y]
            vertices[ii] = vertex
            print("\(vertex.pos)")
        }

        guard let commandBuffer = commandQueue.makeCommandBuffer(),
              let renderPassDescriptor = view.currentRenderPassDescriptor else {
            return
        }
        let color = MTLClearColor(red: 0, green: 0, blue: 0, alpha: 1)
        renderPassDescriptor.colorAttachments[0].clearColor = color

        guard let renderEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor) else {
            return
        }


        renderEncoder.setRenderPipelineState(pipelineState)
        renderEncoder.setVertexBuffer(vertexBuffer, offset: 0, index: 0)
        renderEncoder.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 3)

        renderEncoder.endEncoding()
        guard let drawable = view.currentDrawable else {
            return
        }

        commandBuffer.present(drawable)
        commandBuffer.commit()
    }
}

File: dotnet/MetalTest/Renderer.cs
using System;
using System.Linq;
using System.Runtime.InteropServices;
using CoreGraphics;
using Metal;
using MetalKit;
using OpenTK;

namespace MetalTest
{
    public static class FloatExtensions
    {
        public static byte[] ToBytes(this float[] floats)
        {
            return floats.SelectMany(BitConverter.GetBytes)
                .ToArray();
        }
    }

    public struct Vertex
    {
        public Vector4 Color { get; }
        public Vector2 Pos { get; }

        public Vertex(Vector4 color, Vector2 pos)
        {
            Color = color;
            Pos = pos;
        }
    }

    public class Renderer : MTKViewDelegate
    {
        private readonly IMTLDevice _device;
        private readonly IMTLCommandQueue _commandQueue;
        private readonly IMTLRenderPipelineState _pipelineState;
        private readonly IMTLBuffer _vertexBuffer;

        public Renderer(MTKView view)
        {
            _device = view.Device;
            _commandQueue = _device.CreateCommandQueue();
            _pipelineState = BuildRenderPipeline(view.Device, view);

            var vertices = new[]
            {
                new Vertex(new Vector4(1, 0, 0, 1), new Vector2(-1, -1)),
                new Vertex(new Vector4(0, 1, 0, 1), new Vector2(0, 1)),
                new Vertex(new Vector4(0, 0, 1, 1), new Vector2(1, -1)),
            };
            _vertexBuffer = _device.CreateBuffer(vertices, 0);
            
            var rawsize = Marshal.SizeOf<Vertex>();
            for (var ii = 0; ii < vertices.Length; ii++)
            {
                var vertex = vertices[ii];
                var rawdata = new byte[rawsize];
                var pinnedUniforms = GCHandle.Alloc(vertex, GCHandleType.Pinned);
                var ptr = pinnedUniforms.AddrOfPinnedObject();
                Marshal.Copy(ptr, rawdata, 0, rawsize);
                pinnedUniforms.Free();

                Marshal.Copy(rawdata, 0, _vertexBuffer.Contents + rawsize * ii, rawsize);
            }
        }

        private static IMTLRenderPipelineState BuildRenderPipeline(IMTLDevice device, MTKView view)
        {
            var library = device.CreateDefaultLibrary();
            var pipelineDescriptor = new MTLRenderPipelineDescriptor
            {
                VertexFunction = library.CreateFunction("vertexShader"),
                FragmentFunction = library.CreateFunction("fragmentShader")
            };
            pipelineDescriptor.ColorAttachments[0].PixelFormat = view.ColorPixelFormat;
            var result = device.CreateRenderPipelineState(pipelineDescriptor, out var error);
            if (error != null)
            {
                throw new Exception($"Building pipeline failed: {error}");
            }
            return result;
        }

        public override void Draw(MTKView view)
        {
            var renderPassDescriptor = view.CurrentRenderPassDescriptor;
            renderPassDescriptor.ColorAttachments[0].ClearColor = new MTLClearColor(1, 0, 0, 1);
            var commandBuffer = _commandQueue.CommandBuffer();
            var renderEncoder = commandBuffer.CreateRenderCommandEncoder(renderPassDescriptor);
            // more commands to come

            renderEncoder.SetRenderPipelineState(_pipelineState);
            renderEncoder.SetVertexBuffer(_vertexBuffer, 0, 0);
            renderEncoder.DrawPrimitives(MTLPrimitiveType.Triangle, 0, 3);

            // Finalizes encoding of drawing commands. Does _not_ send commands to GPU.
            renderEncoder.EndEncoding();
            // Tell Metal to send the rendering result to the MTKView when rendering completes
            commandBuffer.PresentDrawable(view.CurrentDrawable);

            // Queues qcommands to GPU.
            commandBuffer.Commit();
        }

        public override void DrawableSizeWillChange(MTKView view, CGSize size)
        {

        }
    }
}
